File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/V3ErrorStatusAdapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import noie.linmimeng.noiedigitalsystem.api.proto.common.v1.ErrorCategory;
import noie.linmimeng.noiedigitalsystem.api.proto.common.v1.ErrorStatus;
import noie.linmimeng.noiedigitalsystem.api.result.NdsError;

/**
 * v3 ErrorStatus Adapter - NdsError ↔ v3 ErrorStatus.
 *
 * @since 3.0.0
 */
public final class V3ErrorStatusAdapter {

    private V3ErrorStatusAdapter() {
        // [Index] NDS-JAVA-V3-ERRORSTATUS-001 [Constraint] Utility class; instantiation is prohibited.
    }

    public static ErrorStatus toProto(NdsError error, ErrorCategory category, Integer retryAfterSeconds) {
        Objects.requireNonNull(error, "error");
        Objects.requireNonNull(category, "category");

        ErrorStatus.Builder builder = ErrorStatus.newBuilder()
            .setCode(error.code())
            .setMessage(error.message())
            .setCategory(category);

        if (retryAfterSeconds != null && retryAfterSeconds >= 0) {
            builder.setRetryAfterSeconds(retryAfterSeconds);
        }

        if (error.details() != null) {
            Map<String, String> details = new HashMap<>();
            for (Map.Entry<String, Object> entry : error.details().entrySet()) {
                details.put(entry.getKey(), String.valueOf(entry.getValue()));
            }
            builder.putAllDetails(details);
        }

        return builder.build();
    }

    public static NdsError fromProto(ErrorStatus status) {
        Objects.requireNonNull(status, "status");

        Map<String, Object> details = new HashMap<>();
        details.putAll(status.getDetailsMap());
        details.put("error_category", status.getCategory().name());
        if (status.hasRetryAfterSeconds()) {
            details.put("retry_after_seconds", status.getRetryAfterSeconds());
        }

        return NdsError.of(status.getCode(), status.getMessage(), details);
    }
}

File: csharp/src/Noie.Nds.Api/Adapter/V3ErrorStatusAdapter.cs
using Nds.Common.V1;
using Noie.Nds.Api.Result;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-V3-ERRORSTATUSADAPTER-000] v3 error status adapter.
///
/// <para>[Semantic] Convert between domain <see cref="NdsError"/> and v3 <see cref="ErrorStatus"/>.</para>
/// </summary>
public static class V3ErrorStatusAdapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-V3-ERRORSTATUSADAPTER-010] Convert domain error to v3 ErrorStatus.
    /// </summary>
    public static ErrorStatus ToProto(NdsError error, ErrorCategory category, int? retryAfterSeconds = null)
    {
        ArgumentNullException.ThrowIfNull(error);

        var status = new ErrorStatus
        {
            Code = error.Code,
            Message = error.Message,
            Category = category
        };

        if (retryAfterSeconds is >= 0)
            status.RetryAfterSeconds = retryAfterSeconds.Value;

        foreach (var (key, value) in error.Details)
            status.Details[key] = value?.ToString() ?? string.Empty;

        return status;
    }

    /// <summary>
    /// [Index: NDS-CSHARP-V3-ERRORSTATUSADAPTER-011] Convert v3 ErrorStatus to domain error.
    /// </summary>
    public static NdsError FromProto(ErrorStatus status)
    {
        ArgumentNullException.ThrowIfNull(status);

        var details = new Dictionary<string, object>();
        foreach (var (key, value) in status.Details)
            details[key] = value;

        // Category/retryAfter are part of v3 semantics; preserve as structured details.
        details["error_category"] = status.Category.ToString();
        if (status.RetryAfterSeconds != 0)
            details["retry_after_seconds"] = status.RetryAfterSeconds;

        return NdsError.Of(status.Code, status.Message, details);
    }
}

